diff --git a/packages/flutter_tools/lib/src/aop/aspectd.dart b/packages/flutter_tools/lib/src/aop/aspectd.dart
new file mode 100644
index 00000000000..17292b5eac0
--- /dev/null
+++ b/packages/flutter_tools/lib/src/aop/aspectd.dart
@@ -0,0 +1,133 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:crypto/crypto.dart';
+import 'package:yaml/yaml.dart';
+import 'package:package_config/package_config.dart';
+
+import '../artifacts.dart';
+import '../base/common.dart';
+import '../base/file_system.dart';
+import '../dart/package_map.dart';
+import '../globals.dart' as globals;
+
+const String frontendServerDartSnapshot = 'frontend_server_aot.dart.snapshot';
+const String sYamlConfigName = 'aop_config.yaml';
+const String key_flutter_tools_hook = 'flutter_tools_hook';
+const String key_project_name = 'project_name';
+const String key_exec_path = 'exec_path';
+const String beike_aspectd = 'beike_aspectd';
+const String inner_path = 'inner';
+const String globalPackagesPath = '.dart_tool/package_config.json';
+
+class AspectdHook {
+  static Future<Directory?> getPackagePathFromConfig(
+      String packageConfigPath, String packageName) async {
+    final PackageConfig packageConfig = await loadPackageConfigWithLogging(
+      globals.fs.file(packageConfigPath),
+      logger: globals.logger,
+    );
+    if (packageConfig.packages.isNotEmpty) {
+      final Iterable<Package> matches = packageConfig.packages.where(
+          (Package element) => element.name == packageName);
+      if (matches.isEmpty) {
+        return null;
+      }
+      return globals.fs.directory(matches.first.root.toFilePath());
+    }
+    return null;
+  }
+
+  static Future<Directory?> getFlutterFrontendServerDirectory(
+      String packagesPath) async {
+    final Directory? directory =
+        await getPackagePathFromConfig(packagesPath, beike_aspectd);
+
+    if (directory == null) {
+      return null;
+    }
+
+    return globals.fs.directory(globals.fs.path
+        .join(directory.absolute.path, inner_path, 'flutter_frontend_server'));
+  }
+
+  static bool configFileExists() {
+    final String configYamlPath =
+        globals.fs.path.join(globals.fs.currentDirectory.path, sYamlConfigName);
+
+    if (globals.fs.file(configYamlPath).existsSync()) {
+      final dynamic yamlInfo =
+          loadYaml(globals.fs.file(configYamlPath).readAsStringSync());
+
+      if (yamlInfo == null) {
+        return false;
+      }
+
+      if (yamlInfo[key_flutter_tools_hook] is! YamlList) {
+        return false;
+      }
+
+      final YamlList yamlNodes = yamlInfo[key_flutter_tools_hook] as YamlList;
+
+      if (yamlNodes.nodes.isNotEmpty) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  static Future<void> enableAspectd() async {
+    final Directory currentDirectory = globals.fs.currentDirectory;
+
+    final String packagesPath = globals.fs.path
+        .join(currentDirectory.absolute.path, globalPackagesPath);
+
+    if (!globals.fs.file(packagesPath).existsSync()) {
+      return;
+    }
+
+    final Directory? flutterFrontendServerDirectory =
+        await getFlutterFrontendServerDirectory(packagesPath);
+
+    if (flutterFrontendServerDirectory == null) {
+      return;
+    }
+
+    await checkAspectdFlutterFrontendServerSnapshot(
+        flutterFrontendServerDirectory);
+  }
+
+  static Future<void> checkAspectdFlutterFrontendServerSnapshot(
+      Directory flutterFrontendServerDirectory) async {
+    final String aspectdFlutterFrontendServerSnapshot = globals.fs.path.join(
+        flutterFrontendServerDirectory.absolute.path,
+        frontendServerDartSnapshot);
+    final String? defaultFlutterFrontendServerSnapshot = globals.artifacts
+        ?.getArtifactPath(Artifact.frontendServerSnapshotForEngineDartSdk);
+    final File defaultServerFile =
+        globals.fs.file(defaultFlutterFrontendServerSnapshot);
+    final File aspectdServerFile =
+        globals.fs.file(aspectdFlutterFrontendServerSnapshot);
+
+    if (!aspectdServerFile.existsSync()) {
+      return;
+    }
+
+    if (defaultServerFile.existsSync()) {
+      if (md5.convert(defaultServerFile.readAsBytesSync()) ==
+          md5.convert(aspectdServerFile.readAsBytesSync())) {
+        return;
+      }
+
+      globals.fs.file(defaultFlutterFrontendServerSnapshot).deleteSync();
+    }
+
+    aspectdServerFile.copySync(defaultFlutterFrontendServerSnapshot!);
+
+    print('[aop]: New frontend server snapshot updated');
+  }
+}
diff --git a/packages/flutter_tools/lib/src/build_system/targets/common.dart b/packages/flutter_tools/lib/src/build_system/targets/common.dart
index 1be5f6aff5a..5b70446e66d 100644
--- a/packages/flutter_tools/lib/src/build_system/targets/common.dart
+++ b/packages/flutter_tools/lib/src/build_system/targets/common.dart
@@ -24,6 +24,7 @@ import 'dart_plugin_registrant.dart';
 import 'icon_tree_shaker.dart';
 import 'localizations.dart';
 import 'native_assets.dart';
+import '../../aop/aspectd.dart';
 
 /// Copies the pre-built flutter bundle.
 // This is a one-off rule for implementing build bundle in terms of assemble.
@@ -165,6 +166,7 @@ class KernelSnapshot extends Target {
 
   @override
   Future<void> build(Environment environment) async {
+    await AspectdHook.enableAspectd();
     final compiler = KernelCompiler(
       fileSystem: environment.fileSystem,
       logger: environment.logger,
diff --git a/packages/flutter_tools/lib/src/commands/build_bundle.dart b/packages/flutter_tools/lib/src/commands/build_bundle.dart
index e2252f23c99..965dd60209f 100644
--- a/packages/flutter_tools/lib/src/commands/build_bundle.dart
+++ b/packages/flutter_tools/lib/src/commands/build_bundle.dart
@@ -13,6 +13,7 @@ import '../globals.dart' as globals;
 import '../project.dart';
 import '../runner/flutter_command.dart';
 import 'build.dart';
+import '../aop/aspectd.dart';
 
 class BuildBundleCommand extends BuildSubCommand {
   BuildBundleCommand({
@@ -107,6 +108,8 @@ class BuildBundleCommand extends BuildSubCommand {
 
   @override
   Future<FlutterCommandResult> runCommand() async {
+    await AspectdHook.enableAspectd();
+
     final String targetPlatform = stringArg('target-platform')!;
     final TargetPlatform platform = getTargetPlatformForName(targetPlatform);
     // Check for target platforms that are only allowed via feature flags.
diff --git a/packages/flutter_tools/lib/src/compile.dart b/packages/flutter_tools/lib/src/compile.dart
index ca04cfb345a..4a0390b8b50 100644
--- a/packages/flutter_tools/lib/src/compile.dart
+++ b/packages/flutter_tools/lib/src/compile.dart
@@ -18,6 +18,7 @@ import 'base/logger.dart';
 import 'base/platform.dart';
 import 'build_info.dart';
 import 'convert.dart';
+import 'aop/aspectd.dart';
 
 /// Opt-in changes to the dart compilers.
 const kDartCompilerExperiments = <String>[];
@@ -367,6 +368,10 @@ class KernelCompiler {
             '-Dflutter.dart_plugin_registrant=$dartPluginRegistrantUri',
           ],
           if (nativeAssets != null) ...<String>['--native-assets', nativeAssets],
+          if (AspectdHook.configFileExists()) ...<String>[
+            '--aop',
+            '1',
+          ],
           // See: https://github.com/flutter/flutter/issues/103994
           '--verbosity=error',
           ...?extraFrontEndOptions,
@@ -878,6 +883,10 @@ class DefaultResidentCompiler implements ResidentCompiler {
           if (nativeAssetsUri != null) ...<String>['--native-assets', nativeAssetsUri],
           if (platformDill != null) ...<String>['--platform', platformDill!],
           if (unsafePackageSerialization) '--unsafe-package-serialization',
+          if (AspectdHook.configFileExists()) ...<String>[
+            '--aop',
+            '1',
+          ],
           // See: https://github.com/flutter/flutter/issues/103994
           '--verbosity=error',
           ...?extraFrontEndOptions,
